### ISSUE TITLE
No error string on failure to change egid

### ISSUE THREAD

**Issue by shtrom** [munin-monitoring/munin#1204]
**Describe the bug**
For some reason at the moment, my system fails to setegid in munin-node. Regardless of the cause, when it fails to do so, and empty error message is shown:

```
2019/06/16-19:00:07 [24336]     2019/06/16-19:00:07 [24407] # FATAL: Plugin 'systemd_units' Can't drop privileges: at /usr/share/perl5/vendor_perl/Munin/Node/Service.pm line 228.\n.
```
I found through trial and error that quoting the `$!` in the `croak` line in `_set_xid` (line 246) fixed the error-display issue:
```sub _set_xid {
    my ($class, $x, $id) = @_;

    # According to pervar manpage, assigning to $<, $> etc results in
    # a system call. So we need to check $! for errors.
    $! = undef;
    $$x = $id;
    croak "$!' if $!;
}
```

**Desktop (please complete the following information):**
 - OS+Distribution Version: ArchLinux x86_64
 - Munin Version [e.g. 2.0.35] 2.0.47-2
- Perl: 5.30.0-2

**Comment by sumpfralle** [comment]
Yes, indeed Perl 5.30.0 seems to have bug with regard to the handling of `$EFFECTIVE_GROUP_ID` (see [#134194](https://rt.perl.org/Ticket/Display.html?id=134194)).

I would advise to comment out line 228 in `Munin/Node/Service.pm`:
```perl
Munin::Node::OS->set_effective_group_id($egids) unless $rgid == $root_gid;
```

See #1202 for the related discussion as a munin issue.

Regarding an improved logging message: I think, the current message is quite suitable.
But maybe you have a suggestion?

**Comment by shtrom** [comment]
On Sun 16 Jun 2019 at 06:07:09 -0700, Lars Kruse wrote:
>Yes, indeed Perl 5.30.0 seems to have bug with regard to the handling of `$EFFECTIVE_GROUP_ID` (see [#134194](https://rt.perl.org/Ticket/Display.html?id=134194)).
>
>I would advise to comment out line 228 in `Munin/Node/Service.pm`:
>```perl
>Munin::Node::OS->set_effective_group_id($egids) unless $rgid == $root_gid;
>```

I guess you mean as an ad hoc fix waiting for the Perl fix? I ended up 
downgrading Perl and waiting.

>See #1202 for the related discussion as a munin issue.

Yup, I tracked all this down. It seems they have an upstream fix in, 
though, so it shouldn't be too long.

>Regarding an improved logging message: I think, the current message is quite suitable.
>But maybe you have a suggestion?

Well, it didn't seem to work very well: I got the line where the error 
occured, but not the actual error message about an invalid operation. 

Simply putting the `$!` in quotes for `print` purposes fixed that issue, 
which is what I am suggesting here.

-- 
Olivier Mehani <shtrom@ssji.net>
PGP fingerprint: 4435 CF6A 7C8D DD9B E2DE  F5F9 F012 A6E2 98C6 6655
Confidentiality cannot be guaranteed on emails sent or received unencrypted.

**Comment by sumpfralle** [comment]
Thank you for the details!

Just for clarity: would the following change be in line with your thoughts?

```patch
- logger("# FATAL: Plugin '$service' Can't drop privileges: $EVAL_ERROR.");
+ logger("# FATAL: Plugin '$service' Can't drop privileges: $EVAL_ERROR ($!).");
```

**Comment by shtrom** [comment]
On Mon 17 Jun 2019 at 15:37:22 -0700, Lars Kruse wrote:
>Just for clarity: would the following change be in line with your thoughts?
>
>```patch
>- logger("# FATAL: Plugin '$service' Can't drop privileges: $EVAL_ERROR.");
>+ logger("# FATAL: Plugin '$service' Can't drop privileges: $EVAL_ERROR ($!).");
>```

Not quite, as `$EVAL_ERROR` should already capture `$!`.

The problem I saw was in OS.pm, in `_set_xid` (speaking from memory): it 
has a line that does

```
croak $! if $!;
```

and I found I didn't actually report anything in `$!` unless it was 
quoted:

```
croak "$!" if $!;
```

-- 
Olivier Mehani <shtrom@ssji.net>
PGP fingerprint: 4435 CF6A 7C8D DD9B E2DE  F5F9 F012 A6E2 98C6 6655
Confidentiality cannot be guaranteed on emails sent or received unencrypted.

**Comment by sumpfralle** [comment]
Ah - thank you for the explanation.

Fixed in `stable-2.0`.
